policy/caa.go

```diff
@@ -0,0 +1,282 @@
+					if record.Header().Rrtype == dns.TypeSOA {
+						r, err := pubU.Resolve(record.(*dns.SOA).Hdr.Name, dns.TypeNS, dns.ClassINET)
+						if err != nil {
+							return "", "", fmt.Errorf("Resolving '%s' failed: %s", record.(*dns.SOA).Hdr.Name, err)
+						}
+						if r.HaveData {
+							for _, nsRecord := range r.Rr {
+								if nsRecord.Header().Rrtype == dns.TypeNS {
+									// just grab the first NS domain
+									nsName = nsRecord.(*dns.NS).Ns
+									rootDomain = nsRecord.(*dns.NS).Hdr.Name
+									break
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	if nsName == "" || rootDomain == "" {
+		return "", "", fmt.Errorf("Couldn't retrieve authoritative nameserver or zone root for '%s'", domain)
+	}
+	rootDomain = strings.TrimRight(rootDomain, ".")
+
+	nsIPs, err := pubU.LookupIP(nsName)
+	if err != nil {
+		return "", "", fmt.Errorf("Resolving '%s' failed: %s", nsName, err)
+	}
+	if len(nsIPs) > 0 {
+		// return first IP for NS
+		return nsIPs[0].String(), rootDomain, nil
+	}
+	return "", "", fmt.Errorf("Address lookup did not return any IPs for '%s'", nsName)
+}
+
+func getDNSKeys(pubU *unbound.Unbound, authU *unbound.Unbound, domain string, rootDomain string) (bool, error) {
+	splitDomain := strings.Split(rootDomain, ".")
+	// build Trust Anchor list
+	var taKeys []string
+	for i := range splitDomain {
+		taDomain := strings.Join(splitDomain[i:], ".")
+		r, err := pubU.Resolve(dns.Fqdn(taDomain), dns.TypeDNSKEY, dns.ClassINET)
+		if err != nil {
+			return false, fmt.Errorf("Resolving '%s' failed: %s", taDomain, err)
+		}
+		if r.HaveData {
+			for _, ta := range r.AnswerPacket.Answer {
+				if ta.Header().Rrtype == dns.TypeDNSKEY {
+					taKeys = append(taKeys, ta.String())
+				}
+			}
+		} else {
+			// cannot complete chain of trust, abort without error
+			return false, nil
+		}
+	}
+
+	// preload root key since every trust chain requires it
+	taKeys = append(taKeys, rootDNSKey)
+	for _, key := range taKeys {
+		if err := authU.AddTa(key); err != nil {
+			// only build *Chain of trust* if we have keys to make a full chain
+			// from the root domain to . (ie [google.com. -> com. -> .])
+			return false, fmt.Errorf("Couldn't add Trust Anchor: %s", err)
+		}
+	}
+
+	return true, nil
+}
+
+func getCaa(u *unbound.Unbound, domain string, alias bool) ([]*CAA, error) {
+	var CAAs []*CAA
+	if alias {
+		canonName, err := u.LookupCNAME(dns.Fqdn(domain))
+		if err != nil {
+			return CAAs, fmt.Errorf("CNAME lookup for '%s' failed: %s", domain, err)
+		}
+		if canonName == "" || canonName == domain {
+			return CAAs, nil
+		}
+		domain = canonName
+	}
+
+	// finally query for the CAA records
+	r, err := u.Resolve(dns.Fqdn(domain), dns.TypeCAA, dns.ClassINET)
+	if err != nil {
+		return CAAs, fmt.Errorf("Resolving '%s' failed: %s", domain, err)
+	}
+
+	// check if response is bogus
+	if !r.Bogus {
+		if r.HaveData {
+			for _, caaRecord := range r.Data {
+				CAAs = append(CAAs, newCAA(caaRecord))
```

Comment:
It seems like this is the only situation in which CAAs will be non-empty. I think it makes more sense to move the declaration / initialization of CAAs down to just inside the `if r.HaveData` clause, and then return after exiting the for loop. That way we can explicitly return an empty slice in all other cases.